yarn-project/sequencer-client/src/block_builder/circuit_block_builder.ts

```diff
@@ -467,48 +467,84 @@ export class CircuitBlockBuilder implements BlockBuilder {
    * @returns
    */
   public async performBaseRollupBatchInsertionProofs(leaves: Buffer[]): Promise<LowNullifierWitnessData[] | undefined> {
-    // Keep track of the touched during batch insertion
-    const touchedNodes: Set<number> = new Set<number>();
+    // Keep track of touched low nullifiers
+    const touched = new Map<number, bigint[]>();
 
-    // Return data
+    // Accumulators
     const lowNullifierWitnesses: LowNullifierWitnessData[] = [];
+    const pendingInsertionSubtree: NullifierLeafPreimage[] = [];
+
+    // Start info
     const dbInfo = await this.db.getTreeInfo(MerkleTreeId.NULLIFIER_TREE);
     const startInsertionIndex: bigint = dbInfo.size;
-    let currInsertionIndex: bigint = startInsertionIndex;
 
-    // Leaf data of hte leaves to be inserted
-    const insertionSubtree: NullifierLeafPreimage[] = [];
+    // Prefill empty values
+    const zeroPreimage: NullifierLeafPreimage = new NullifierLeafPreimage(new Fr(0n), new Fr(0n), 0);
+    const emptySP = new SiblingPath();
+    emptySP.data = Array(dbInfo.depth).fill(
+      Buffer.from('0000000000000000000000000000000000000000000000000000000000000000', 'hex'),
+    );
+    const emptyWitness: LowNullifierWitnessData = {
+      preimage: NullifierLeafPreimage.empty(),
+      index: 0n,
+      siblingPath: emptySP,
+    };
+
+    // Get insertion path for each leaf
+    for (let i = 0; i < leaves.length; i++) {
+      const newValue = toBigIntBE(leaves[i]);
+
+      // Keep space and just insert zero values
+      if (newValue === 0n) {
+        pendingInsertionSubtree.push(zeroPreimage);
+        lowNullifierWitnesses.push(emptyWitness);
+        continue;
+      }
 
-    // Low nullifier membership proof sibling paths
-    for (const leaf of leaves) {
-      const newValue = toBigIntBE(leaf);
       const indexOfPrevious = await this.db.getPreviousValueIndex(MerkleTreeId.NULLIFIER_TREE, newValue);
 
-      // NOTE: null values for nullfier leaves are being changed to 0n current impl is a hack
-      // Default value
-      const nullifierLeaf: NullifierLeafPreimage = new NullifierLeafPreimage(new Fr(newValue), new Fr(0n), 0);
-      if (touchedNodes.has(indexOfPrevious.index) || newValue === 0n) {
-        // If the node has already been touched, then we return an empty leaf and sibling path
-        const emptySP = new SiblingPath();
-        emptySP.data = Array(dbInfo.depth).fill(
-          Buffer.from('0000000000000000000000000000000000000000000000000000000000000000', 'hex'),
-        );
-        const witness: LowNullifierWitnessData = {
-          preimage: NullifierLeafPreimage.empty(),
-          index: 0n,
-          siblingPath: emptySP,
-        };
-        lowNullifierWitnesses.push(witness);
+      // If a touched node has a value that is less greater than the current value
```

Comment:
nit: expand the comment -> the case where you have a multiple nodes share the same lo nullifier